hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/process/ConsumerProcessSupervisor.java

```diff
@@ -0,0 +1,156 @@
+package pl.allegro.tech.hermes.consumers.supervisor.process;
+
+import org.jctools.queues.MpscArrayQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import pl.allegro.tech.hermes.api.SubscriptionName;
+import pl.allegro.tech.hermes.common.config.ConfigFactory;
+import pl.allegro.tech.hermes.common.config.Configs;
+import pl.allegro.tech.hermes.consumers.consumer.Consumer;
+import pl.allegro.tech.hermes.consumers.supervisor.ConsumersExecutorService;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Future;
+
+public class ConsumerProcessSupervisor implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(ConsumerProcessSupervisor.class);
+
+    private final MpscArrayQueue<Signal> taskQueue = new MpscArrayQueue<>(1000);
+
+    private final RunningConsumerProcesses runningProcesses = new RunningConsumerProcesses();
+
+    private final Retransmitter retransmitter;
+
+    private final ConsumersExecutorService executor;
+
+    private final Clock clock;
+
+    private long unhealthyAfter;
+
+    public ConsumerProcessSupervisor(ConsumersExecutorService executor,
+                                     Retransmitter retransmitter,
+                                     Clock clock,
+                                     ConfigFactory configs) {
+        this.executor = executor;
+        this.retransmitter = retransmitter;
+        this.clock = clock;
+        this.unhealthyAfter = configs.getIntProperty(Configs.CONSUMER_BACKGROUND_SUPERVISOR_UNHEALTHY_AFTER);
+    }
+
+    public void accept(Signal signal) {
+        taskQueue.offer(signal);
+    }
+
+    @Override
+    public void run() {
+        logger.info("Starting process supervisor loop");
+        long currentTime = clock.millis();
+
+        try {
+            checkHealth();
+            taskQueue.drain(this::processSignal);
+        }
+        catch(Exception exception) {
+            logger.error("Process supervisor loop failed", exception);
+        }
+
+        logger.info("Process supervisor loop took {} ms to check all consumers", clock.millis() - currentTime);
+    }
+
+    private void checkHealth() {
```

Comment:
I think that this should be called `restartUnhealthy` because just checking imply that decision what to do is postponed after processing result.